src/number_theory/wilson.lean

```diff
@@ -0,0 +1,49 @@
+/-
+Copyright (c) 2022 John Nicol. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: John Nicol
+-/
+import number_theory.legendre_symbol.gauss_eisenstein_lemmas
+
+/-!
+# Wilson's theorem.
+
+This file contains a proof of Wilson's theorem.
+
+The heavy lifting is mostly done by the previous `wilsons_lemma`,
+but here we also prove the other logical direction.
+
+This could be generalized to similar results about finite abelian groups.
+
+## References
+
+* [Wilson's Theorem](https://en.wikipedia.org/wiki/Wilson%27s_theorem)
+
+-/
+
+open_locale nat
+
+namespace nat
+
+/-- For `n > 1`, `(n-1)!` is congruent to `-1` modulo `n` only if n is prime. --/
+lemma prime_of_fac_equiv_neg_one
+  {n : ℕ} (h : ((n - 1)! : zmod n) = -1) (h1 : 1 < n) : prime n :=
+begin
+  by_contradiction h2,
+  obtain ⟨m, hm1, hm2 : 1 < m, hm3⟩ := exists_dvd_of_not_prime2 h1 h2,
+  rw [lt_iff_add_one_le, nat.add_le_to_le_sub m h1.le] at hm3,
+  replace hm3 := nat.dvd_factorial (pos_of_gt hm2) hm3,
+  refine hm2.ne' (nat.dvd_one.mp ((nat.dvd_add_right hm3).mp (hm1.trans _))),
```

Comment:
You can replace effect of the `rw` line with `le_pred_of_lt`. I also switched this to using `have` rather than `replace`, but you don't have to if you don't want. I also added a type ascription just to keep track of what's going on in the proof, but again you don't have to include it.
```suggestion
  have hm : m ∣ (n - 1)! := nat.dvd_factorial (pos_of_gt hm2) (le_pred_of_lt hm3),
  refine hm2.ne' (nat.dvd_one.mp ((nat.dvd_add_right hm).mp (hm1.trans _))),
```